File: 백준/Silver/1874. 스택 수열/스택 수열.c
#include <stdio.h>
#include <stdbool.h>

#define BUFSIZE 1 << 17
char readbuf[BUFSIZE];
int rp = BUFSIZE;
char ReadChar(){
    if(rp == BUFSIZE){
        fread(readbuf, 1, BUFSIZE, stdin);
        rp = 0;
    }
    return readbuf[rp++];
}
int ReadInt(){
    char c;
    int ret = 0;
    while(c < '0' || c > '9') c = ReadChar();
    while(c >= '0' && c <= '9'){
        ret = ret * 10 + (c & 0xf);
        c = ReadChar();
    }
    return ret;
}


int stack[100001];
char result[400002];
int top = -1;
int idx;

void push(int data){
    stack[++top] = data;
}
int pop(){
    return stack[top--];
}
void Write(char c){
    result[idx++] = c;
    result[idx++] = '\n';
}

int main(void){
    int N, num;
    int k = 1;

    N = ReadInt();
    for(int i = 1; i <= N; i++){
        num = ReadInt();
        while(num >= k){
            push(k++);
            Write('+');
        }
        if(top != -1 && stack[top] == num){
            pop();
            Write('-');
        }
    }
    if(top != -1) printf("NO\n");
    else printf("%s", result);
}

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    private int[] stack = new int[100001];
    private int top = -1;

    public void push(int data){
        stack[++top] = data;
    }
    public int pop(){
        return stack[top--];
    }

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int k = 1;
        int N = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < N; i++){
            int num = Integer.parseInt(br.readLine());
            while(num >= k){
                push(k++);
                sb.append('+');
                sb.append('\n');
            }
            if(top != -1 && stack[top] == num){
                pop();
                sb.append('-');
                sb.append('\n');
            }
        }    
        if(top != -1) {
            System.out.println("NO");
            return;
        } else bw.write(sb.toString());
        
        bw.flush();   
        bw.close();
        br.close();
    }
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}
